File: apps/windows/Finance/FinanceTextImport.cs
using System.Globalization;

namespace Andivum_Windows.Finance;

public static class FinanceTextImport
{
    public static bool TryCreateDraft(
        string text,
        string fileName,
        IReadOnlyCollection<FinanceCategory> categories,
        out FinanceDraft draft)
    {
        draft = null!;
        var lines = text
            .Replace("\r", string.Empty, StringComparison.Ordinal)
            .Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
        var title = lines.FirstOrDefault(line => line.StartsWith("Subject:", StringComparison.OrdinalIgnoreCase))?
            .Split(':', 2)[1].Trim();
        title ??= Path.GetFileNameWithoutExtension(fileName);
        var items = new List<FinanceDraftItem>();
        foreach (var line in lines)
        {
            var fields = line.Contains(';')
                ? line.Split(';', StringSplitOptions.TrimEntries)
                : line.Contains('\t')
                    ? line.Split('\t', StringSplitOptions.TrimEntries)
                    : [line];
            var amountField = fields[^1].Trim();
            decimal amount;
            if (fields.Length == 1)
            {
                var amountMatch = System.Text.RegularExpressions.Regex.Match(
                    line,
                    @"(?<!\d)(\d+(?:[.,]\d{1,2})?)(?!\d)\s*(?:₽|RUB|руб)?$",
                    System.Text.RegularExpressions.RegexOptions.IgnoreCase);
                if (!amountMatch.Success)
                {
                    continue;
                }

                amountField = amountMatch.Groups[1].Value;
                fields = [line[..amountMatch.Index].Trim(' ', '-', ':')];
                if (!decimal.TryParse(
                        amountField.Replace(',', '.'),
                        NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint,
                        CultureInfo.InvariantCulture,
                        out amount))
                {
                    continue;
                }
            }
            else if (!decimal.TryParse(
                         amountField.Replace(',', '.'),
                         NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint,
                         CultureInfo.InvariantCulture,
                         out amount))
            {
                continue;
            }

            if (amount < 0)
            {
                continue;
            }
            var minor = checked((long)decimal.Round(amount * 100m, 0, MidpointRounding.AwayFromZero));
            if (minor <= 0)
            {
                continue;
            }
            var category = fields.Length >= 3 && categories.Any(item =>
                item.Slug.Equals(fields[1], StringComparison.OrdinalIgnoreCase))
                ? fields[1]
                : "other.expense";
            items.Add(new FinanceDraftItem(fields[0], 1m, minor, category));
        }

        if (items.Count == 0)
        {
            return false;
        }

        draft = new FinanceDraft(
            FinanceTransactionType.Expense,
            string.IsNullOrWhiteSpace(title) ? "Импорт" : title,
            DateTimeOffset.Now.ToString("yyyy-MM-dd"),
            "RUB",
            items.Sum(item => item.LineTotalMinor),
            items);
        return true;
    }
}

File: apps/android/app/src/main/java/io/github/andivalsei/andivum/FinanceTextImport.kt
package io.github.andivalsei.andivum

import java.time.LocalDate

object FinanceTextImport {
    private val trailingAmount = Regex("(?i)(?<!\\d)(\\d+(?:[.,]\\d{1,3})?)(?!\\d)\\s*(?:₽|RUB|руб)?$")

    fun tryCreateDraft(
        text: String,
        fileName: String,
        categories: Collection<FinanceCategory>,
    ): FinanceDraft? {
        val lines = text.lines().map { it.trim() }.filter { it.isNotBlank() }
        if (lines.isEmpty()) return null
        val subject = lines.firstOrNull { it.startsWith("Subject:", ignoreCase = true) }
            ?.substringAfter(':')?.trim()
            ?.ifBlank { null }
        val items = buildList {
            lines.forEach { line ->
                val fields = when {
                    ';' in line -> line.split(';').map(String::trim)
                    '\t' in line -> line.split('\t').map(String::trim)
                    else -> emptyList()
                }
                val amountText: String
                val itemName: String
                val requestedCategory: String?
                if (fields.size >= 2) {
                    amountText = fields.last()
                    itemName = fields.first()
                    requestedCategory = fields.getOrNull(1)
                } else {
                    val match = trailingAmount.find(line) ?: return@forEach
                    amountText = match.groupValues[1]
                    itemName = line.substring(0, match.range.first).trim(' ', '-', ':')
                    requestedCategory = null
                }
                val amount = runCatching { FinanceMoney.parseMinorUnits(amountText, "RUB") }.getOrNull()
                    ?: return@forEach
                if (amount <= 0L || itemName.isBlank()) return@forEach
                val category = requestedCategory
                    ?.takeIf { slug -> categories.any { it.slug.equals(slug, ignoreCase = true) && it.type == "expense" } }
                    ?: "other.expense"
                add(FinanceDraftItem(itemName, 1.0, amount, category))
            }
        }
        if (items.isEmpty()) return null
        return FinanceDraft(
            type = FinanceTransactionType.EXPENSE,
            title = subject ?: fileName.substringBeforeLast('.', fileName).ifBlank { "Импорт" },
            occurredOn = LocalDate.now().toString(),
            currency = "RUB",
            totalMinor = items.sumOf { it.lineTotalMinor },
            items = items,
        )
    }
}
